core/src/test/java/org/elasticsearch/index/engine/CombinedDeletionPolicyTests.java

```diff
@@ -79,40 +77,46 @@ public void testKeepCommitsAfterGlobalCheckpoint() throws Exception {
 
-        long firstMaxSeqNo = randomLongBetween(0, Long.MAX_VALUE - 1);
-        long secondMaxSeqNo = randomLongBetween(firstMaxSeqNo + 1, Long.MAX_VALUE);
+        final long maxSeqNo1 = between(1, 1000);
+        final long translogGen1 = between(1, 100);
+        final MockIndexCommit c1 = mockIndexCommit(maxSeqNo1, translogUUID, translogGen1);
+        final long maxSeqNo2 = maxSeqNo1 + between(1, 1000);
+        final long translogGen2 = translogGen1 + between(1, 100);
+        final MockIndexCommit c2 = mockIndexCommit(maxSeqNo2, translogUUID, translogGen2);
+        globalCheckpoint.set(randomLongBetween(0, maxSeqNo2 - 1)); // Keep both c1 and c2.
+        indexPolicy.onCommit(Arrays.asList(c1, c2));
+        final IndexCommit ref1 = indexPolicy.acquireIndexCommit(true);
+        assertThat(ref1, equalTo(c1));
+        expectThrows(UnsupportedOperationException.class, ref1::delete);
+        final IndexCommit ref2 = indexPolicy.acquireIndexCommit(false);
+        assertThat(ref2, equalTo(c2));
+        expectThrows(UnsupportedOperationException.class, ref2::delete);
+        assertThat(translogPolicy.getMinTranslogGenerationForRecovery(), lessThanOrEqualTo(100L));
 
-        long lastTranslogGen = randomNonNegativeLong();
-        final IndexCommit firstCommit = mockIndexCommit(firstMaxSeqNo, translogUUID, randomLongBetween(0, lastTranslogGen));
-        final IndexCommit secondCommit = mockIndexCommit(secondMaxSeqNo, translogUUID, lastTranslogGen);
-        SnapshotDeletionPolicy snapshotDeletionPolicy = new SnapshotDeletionPolicy(indexPolicy);
-
-        snapshotDeletionPolicy.onInit(Arrays.asList(firstCommit));
-        snapshotDeletionPolicy.snapshot();
-        assertThat(snapshotDeletionPolicy.getSnapshots(), contains(firstCommit));
-
-        // SnapshotPolicy prevents the first commit from deleting, but CombinedPolicy does not retain its translog.
-        globalCheckpoint.set(randomLongBetween(secondMaxSeqNo, Long.MAX_VALUE));
-        snapshotDeletionPolicy.onCommit(Arrays.asList(firstCommit, secondCommit));
-        verify(firstCommit, never()).delete();
-        verify(secondCommit, never()).delete();
-        assertThat(translogPolicy.getMinTranslogGenerationForRecovery(), equalTo(lastTranslogGen));
-        assertThat(translogPolicy.getTranslogGenerationOfLastCommit(), equalTo(lastTranslogGen));
+        globalCheckpoint.set(randomLongBetween(maxSeqNo2, Long.MAX_VALUE));
+        indexPolicy.onCommit(Arrays.asList(c1, c2)); // Policy keeps c2 only, but c1 is snapshotted.
+        assertThat(c1.deleteTimes(), equalTo(0));
+        final IndexCommit ref3 = indexPolicy.acquireIndexCommit(true);
+        assertThat(ref3, equalTo(c2));
+        assertThat(translogPolicy.getMinTranslogGenerationForRecovery(), equalTo(translogGen1));
+        indexPolicy.releaseCommit(ref1); // release acquired commit releases translog and commit
+        indexPolicy.onCommit(Arrays.asList(c1, c2)); // Flush new commit deletes c1
```

Comment:
can we also release c2 and see that it is not deleted? I would also appreciate some randomness here - i.e., a series of commits, choose snapshot at random times. Release at random order and check that all is OK.